feat: implement remote port forwarding (reverse tunneling)

serverleader opened the pull request:
Implements remote port forwarding (reverse tunneling) API for Citadel.

  ## What is Remote Port Forwarding?

  Remote port forwarding allows exposing a local service through a remote SSH server.
  When someone connects to a port on the remote server, that connection is forwarded
  back through the SSH tunnel to the client machine.

  ## Implementation

  ### Client API
  - `SSHClient.createRemotePortForward(host:port:handler:)` - Request remote port forward
  - `SSHClient.cancelRemotePortForward(_:)` - Cancel active forward
  - Uses NIOSSH's GlobalRequest.TCPForwardingRequest API
  - Returns `SSHRemotePortForward` with bound port information

  ### Server API
  - `RemotePortForwardDelegate` protocol for handling forward requests
  - `DefaultRemotePortForwardDelegate` implementation with port range restrictions
  - `SSHServer.enableRemotePortForward(withDelegate:)` to activate feature
  - Implements `GlobalRequestDelegate.tcpForwardingRequest()`

serverleader committed:
feat: implement remote port forwarding (reverse tunneling)

Implements remote port forwarding (reverse tunneling) API for Citadel.

  ## What is Remote Port Forwarding?

  Remote port forwarding allows exposing a local service through a remote SSH server.
  When someone connects to a port on the remote server, that connection is forwarded
  back through the SSH tunnel to the client machine.

  ## Implementation

  ### Client API
  - `SSHClient.createRemotePortForward(host:port:handler:)` - Request remote port forward
  - `SSHClient.cancelRemotePortForward(_:)` - Cancel active forward
  - Uses NIOSSH's GlobalRequest.TCPForwardingRequest API
  - Returns `SSHRemotePortForward` with bound port information

  ### Server API
  - `RemotePortForwardDelegate` protocol for handling forward requests
  - `DefaultRemotePortForwardDelegate` implementation with port range restrictions
  - `SSHServer.enableRemotePortForward(withDelegate:)` to activate feature
  - Implements `GlobalRequestDelegate.tcpForwardingRequest()`

Joannis reviewed (commented):
Awesome work! The NSLog issue needs to be replaced with a structured swift-log approach. Other than that it looks great!

I also have an idea on how to expand the remote port forwarding APIs, but that can and should be a separate PR.

Do you want me to fix the logging, or do you want to take a stab at it yourself?

serverleader committed:
Replaced NSLog with Logger

serverleader committed:
new logger for inbound channel

No need to access client.logger through the weak reference.
Logger is captured once and used safely in the @Sendable closure

serverleader committed:
feat: add high-level NIOAsyncChannel APIs for remote port forwarding

Adds modern async/await-based APIs for both client and server-side remote
port forwarding, replacing low-level EventLoopFuture patterns with ergonomic
NIOAsyncChannel interfaces.

 ## Client-Side Enhancements

  ### High-Level Convenience API
  - `withRemotePortForward(forwardingTo:)` - Simplest API for forwarding to local services
  - Automatically handles bidirectional forwarding using GlueHandler
  - Example: Forward remote port 8080 to local HTTP server on port 3000

  ### NIOAsyncChannel API
  - `withRemotePortForward(configure:onAccept:)` - Generic async/await API
  - Supports custom channel pipeline configuration
  - Type-safe with generic Inbound/Outbound types
  - Uses `channel.executeThenClose { inbound, outbound in }` pattern

  ### Logger Migration
  - Replaced NSLog with structured logging (Logger from swift-log)
  - Uses proper log levels (trace, debug, info, error, warning)
  - Added metadata for better observability
  - Created dedicated logger for inbound channel handling

  ## Server-Side Enhancements

  ### AsyncRemotePortForwardDelegate
  - High-level RemotePortForwardDelegate implementation using NIOAsyncChannel
  - Thread-safe channel storage using Swift actors
  - Configurable host and port whitelisting for security
  - Automatic lifecycle management (tracks and closes server channels)
  - Structured logging throughout

serverleader commented on Examples/RemotePortForwardExample/Sources/RemotePortForwardExample/main.swift:
```diff
@@ -0,0 +1,94 @@
+struct RemotePortForwardExample {
+    static func main() async throws {
+        // Get SSH credentials from environment or use defaults
+        let host = ProcessInfo.processInfo.environment["SSH_HOST"] ?? "localhost"
```
changed the example using it now :)

https://github.com/serverleader/Citadel/commit/448cbbe088c6aeb0e5d08e505cf4e9aafd7722e6
